### PR TITLE
[WIP] Update TKE Driver

### PR DESCRIPTION
**Pull request by GGGitBoy**
Tencent has removed V2 API from their service endpoints. So we need to upgrade TKE Driver to be compatible with the V3 API

Related PR:

https://github.com/rancher/ui/pull/3970

Related Issue:

https://github.com/rancher/rancher/issues/27121

### CODE REVIEW

File: driver/tke_driver.go [hunk 1]
```diff
@@ -198,7 +200,12 @@ func (d *Driver) GetDriverCreateOptions(ctx context.Context) (*types.DriverFlags
 	}
 	driverFlag.Options["zoneId"] = &types.Flag{
 		Type:  types.StringType,
-		Usage: "The zone id of the cluster",
+		Usage: "The zone id of the CVM Instance",
+	}
+	driverFlag.Options["imageId"] = &types.Flag{
+		Type:    types.StringType,
+		Usage:   "The image id of the cluster",
+		Default: &types.Default{DefaultString: "img-pyqx34y1"},
```
Comment (gitlawr) [inline driver/tke_driver.go]: What is a cluster image, is it the image for nodes?
Where does the default value come from? Does it depend on regions?

File: driver/tke_driver.go [hunk 2]
```diff
@@ -125,6 +130,8 @@ type state struct {
 	ClusterInfo types.ClusterInfo
 	// The total number of worker node
 	NodeCount int64
+
+	CreateClusterRequest *tke.CreateClusterRequest
```
Comment (gitlawr) [inline driver/tke_driver.go]: Can we only add what needs to be updated here and avoid using TKE defined object directly in the state.
1. To decouple from specific TKE API/SDK version
2. It overlaps the rest of the state structure a lot

File: driver/tke_driver.go [hunk 3]
```diff
@@ -621,129 +651,110 @@ func (d *Driver) Update(ctx context.Context, info *types.ClusterInfo, opts *type
...
 			log.Infof(ctx, "Scaling down cluster nodes to %d", newState.NodeCount)
-			if err := removeClusterInstances(state, newState, svc, resp); err != nil {
+
+			req, err := removeClusterInstances(state, newState, int64(*nodeCount), resp)
+			if err != nil {
+				return nil, err
+			}
+
+			_, err = svc.DeleteClusterInstances(req)
+			if _, ok := err.(*tcerrors.TencentCloudSDKError); ok {
```
Comment (gitlawr) [inline driver/tke_driver.go]: What if the error is a mismatching type, looks like we don't handle it in this case.

File: driver/tke_driver.go [hunk 4]
```diff
@@ -621,129 +651,110 @@ func (d *Driver) Update(ctx context.Context, info *types.ClusterInfo, opts *type
...
+		log.Infof(ctx, "Updating cluster %s attributes to name: %s, desc: %s", *state.CreateClusterRequest.ClusterBasicSettings.ClusterName, *newState.CreateClusterRequest.ClusterBasicSettings.ClusterName, *newState.CreateClusterRequest.ClusterBasicSettings.ClusterDescription)
 		req, err := getWrapModifyClusterAttributesRequest(state, newState)
 		if err != nil {
 			return nil, err
 		}
 
 		// init the TKE client
-		resp, err := svc.ModifyClusterAttributes(req)
+		_, err = svc.ModifyClusterAttribute(req)
 		if _, ok := err.(*tcerrors.TencentCloudSDKError); ok {
```
Comment (gitlawr) [inline driver/tke_driver.go]: What if the error is a mismatching type, looks like we don't handle it in this case.

File: driver/tke_driver.go [hunk 5]
```diff
@@ -865,32 +880,39 @@ func (d *Driver) GetVersion(ctx context.Context, info *types.ClusterInfo) (*type
...
+	req := tke.NewCreateClusterEndpointVipRequest()
+	req.ClusterId = &clusterID
+
+	reqStatus := tke.NewDescribeClusterEndpointVipStatusRequest()
+	reqStatus.ClusterId = &clusterID
+
+	_, err := svc.CreateClusterEndpointVip(req)
+
+	if _, ok := err.(*tcerrors.TencentCloudSDKError); ok {
+		return fmt.Errorf("an API error has returned: %s", err)
```
Comment (gitlawr) [inline driver/tke_driver.go:972]: ditto

File: driver/tke_driver.go [hunk 6]
```diff
@@ -1013,96 +1035,183 @@ func (d *Driver) GetK8SCapabilities(ctx context.Context, opts *types.DriverOptio
 	return capabilities, nil
 }
 
-func getWrapAddClusterInstancesRequest(state, newState *state, nodeCount int64) (*ccs.AddClusterInstancesRequest, error) {
-	logrus.Debugf("invoking get wrap request of AddClusterInstances")
-	// goodsNum is the new nodes count that will be added to the cluster
-	state.GoodsNum = newState.NodeCount - nodeCount
+func removeClusterInstances(state, newState *state, nodeCount int64, instancesResp *tke.DescribeClusterInstancesResponse) (*tke.DeleteClusterInstancesRequest, error) {
```
Comment (gitlawr) [inline driver/tke_driver.go]: Looks like nodeCount is from instancesResp so we don't need 4 params for this function.

**Comment by loganhz** [comment]
为了保证api不变，我们不应该对ui进行改动，所以需要在后端处理数据类型转换

File: driver/tke_driver.go (hunk 1 above)
Comment (GGGitBoy) [reply driver/tke_driver.go]: It is the image for nodes. The imageId `img-pyqx34y1` corresponds to `ubuntu16.04.1 LTSx86_64`, it is the same as the default value of osName

File: tencentcloud/ccs/models.go [hunk 7]
```diff
@@ -6,222 +6,231 @@ import (
 	tchttp "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common/http"
 )
 
-// CreateClusterRequest defines create cluster request parameters
-type CreateClusterRequest struct {
+//ModifyClusterAttributeRequest defines modify cluster attribute request
```
Comment (gitlawr) [inline tencentcloud/ccs/models.go]: What's the reason not using the models from the SDK? https://github.com/TencentCloud/tencentcloud-sdk-go/blob/17112f5f9a7e2ebea6df4af5407618cbc1d81b5d/tencentcloud/tke/v20180525/models.go#L1852

File: driver/tke_driver.go (hunk 1 above)
Comment (gitlawr) [reply driver/tke_driver.go]: Is there a newer version available? e.g., 20.04

File: driver/tke_driver.go [hunk 8]
```diff
@@ -482,15 +492,14 @@ func (d *Driver) Create(ctx context.Context, opts *types.DriverOptions, _ *types
 	defer storeState(info, state)
 
 	// init tke client and make create cluster api request
-	resp, err := svc.CreateCluster(req, state.EmptyCluster)
-	if _, ok := err.(*tcerrors.TencentCloudSDKError); ok {
+	resp, err := svc.CreateCluster(req)
+
```
Comment (gitlawr) [inline driver/tke_driver.go]: nit: remove this newline

File: driver/tke_driver.go [hunk 9]
```diff
@@ -534,43 +617,39 @@ func waitTKECluster(ctx context.Context, svc *ccs.Client, state *state) error {
 				return err
 			}
 
-			if cluster.CodeDesc != lastMsg {
-				log.Infof(ctx, "provisioning cluster %s: %s", state.ClusterName, cluster.CodeDesc)
-				lastMsg = cluster.CodeDesc
-			}
```
Comment (gitlawr) [inline driver/tke_driver.go]: Are we not able to see provisioning messages now?

File: driver/tke_driver.go [hunk 10]
```diff
@@ -667,12 +753,12 @@ func (d *Driver) Update(ctx context.Context, info *types.ClusterInfo, opts *type
 		}
 
 		// init the TKE client
-		resp, err := svc.ModifyClusterAttributes(req)
-		if _, ok := err.(*tcerrors.TencentCloudSDKError); ok {
+		_, err = svc.ModifyClusterAttribute(req)
+		if err != nil {
 			return nil, err
 		}
 		if err == nil {
```
Comment (gitlawr) [inline driver/tke_driver.go]: Looks like this check is not necessary now.

File: driver/tke_driver.go [hunk 11]
```diff
@@ -865,32 +946,39 @@ func (d *Driver) GetVersion(ctx context.Context, info *types.ClusterInfo) (*type
...
-	req.Operation = operation
+	req := ccs.NewCreateClusterEndpointVipRequest()
+	req.ClusterID = &clusterID
+
+	reqStatus := ccs.NewDescribeClusterEndpointVipStatusRequest()
+	reqStatus.ClusterID = &clusterID
+
+	_, err := svc.CreateClusterEndpointVip(req)
+
+	if err != nil {
```
Comment (gitlawr) [inline driver/tke_driver.go]: Use this common pattern: `if  _, err := svc.CreateClusterEndpointVip(req);err!=nil{}`

File: driver/tke_driver.go [hunk 12]
```diff
@@ -1095,14 +1181,71 @@ func getWrapModifyProjectIDRequest(state, newState *state) (*ccs.ModifyProjectID
...
+func getBoolean(value int64) bool {
+	var boolean bool
+	if value == 1 {
+		boolean = true
+	} else {
+		boolean = false
+	}
+
+	return boolean
+}
```
Comment (gitlawr) [inline driver/tke_driver.go]: {return value == 1} might be better

File: driver/tke_driver.go [hunk 13]
```diff
@@ -1095,14 +1181,71 @@ func getWrapModifyProjectIDRequest(state, newState *state) (*ccs.ModifyProjectID
...
+	svc, err := getTKEServiceClient(state, "GET")
 	if err != nil {
-		return nil, err
+		return zone, err
 	}
-	request := ccs.NewModifyProjectIDRequest()
-	err = request.FromJSONString(string(content))
+
+	request := ccs.NewDescribeZonesRequest()
+	// 通过client对象调用想要访问的接口，需要传入请求对象
```
Comment (gitlawr) [inline driver/tke_driver.go]: This comment does not help

File: driver/tke_driver.go [hunk 14]
```diff
@@ -1095,14 +1181,71 @@ func getWrapModifyProjectIDRequest(state, newState *state) (*ccs.ModifyProjectID
...
+	response, err := svc.DescribeZones(request)
+
```
Comment (gitlawr) [inline driver/tke_driver.go]: remove this newline

File: driver/tke_driver.go [hunk 15]
```diff
@@ -7,29 +7,33 @@ import (
...
+	successStatus  = "Created"
+	failedStatus   = "CreateFailed"
+	notReadyStatus = "ClusterNotReadyError"
+	retries        = 5
+	pollInterval   = 30
+)
+
+var (
+	terminate      = "terminate"
+	managedCluster = "MANAGED_CLUSTER"
```
Comment (gitlawr) [inline driver/tke_driver.go]: Rename the variables to `instanceDeleteModeTerminate` and `managedClusterType` to help people know what they are used for.

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**